src/index.ts

```diff
@@ -1,147 +1,172 @@
     /**
-     * Your user's unique id in your system
+     * A unique identifier for the event
      */
-    external_id?: string;
+    name: string;
     /**
-     * Your user's email address
+     * Any data pertinent to the event
      */
-    email?: string;
+    data: Data;
     /**
-     * Your user's phone number
+     * Any user data associated with the event
+     * All fields ending in "_id" will be used to attribute the event to a particular user
      */
-    phone?: string;
-    [key: string]: any;
-  };
-  /**
-   * A specific event schema version
-   * (optional)
-   */
-  v?: string;
+    user?: {
+      /**
+       * Your user's unique id in your system
+       */
+      external_id?: string;
+      /**
+       * Your user's email address
+       */
+      email?: string;
+      /**
+       * Your user's phone number
+       */
+      phone?: string;
+      [key: string]: any;
+    };
+    /**
+     * A specific event schema version
+     * (optional)
+     */
+    v?: string;
+    /**
+     * An integer representing the milliseconds since the unix epoch at which this event occurred.
+     * Defaults to the current time.
+     * (optional)
+     */
+    ts?: number;
+  }
+
   /**
-   * An integer representing the milliseconds since the unix epoch at which this event occurred.
-   * Defaults to the current time.
-   * (optional)
+   * A set of options for configuring the Inngest client
    */
-  ts?: number;
-}
+  export interface ClientOptions {
+    /**
+     * The base Inngest Source API URL to append the Source API Key to.
+     * Defaults to https://inn.gs/
+     */
+    inngestBaseUrl?: string;
+  }
 
-/**
- * A set of options for configuring the Inngest client
- */
-export interface InngestClientOptions {
   /**
-   * The base Inngest Source API URL to append the Source API Key to.
-   * Defaults to https://inn.gs/e/
+   * A set of options for configuring an Inngest function.
    */
-  inngestApiUrl?: string;
+  export interface FunctionOptions {
+    name: string;
+  }
+
+  export type Steps<Events extends Record<string, any>> = Record<
+    string,
+    InngestStep<Events, any[], any>
+  >;
 }
 
 /**
  * A client for the Inngest Source API
  */
-class Inngest {
+export class Inngest<Events extends Record<string, any>> {
+  public readonly name: string;
+
   /**
    * Inngest Source API Key
    */
-  private apiKey: string;
+  private readonly apiKey: string;
 
   /**
    * Full URL for the Inngest Source API
    */
-  private inngestApiUrl: string;
+  private readonly inngestBaseUrl: URL;
+  private readonly inngestApiUrl: URL;
+  private readonly inngestRegisterUrl: URL;
 
   /**
    * Axios configuration for sending events to Inngest
    */
-  private axiosConfig: AxiosRequestConfig;
+  private readonly client: AxiosInstance;
+
+  private readonly fns: Record<string, InngestFunction<Events>> = {};
 
   /**
    * @param apiKey - An API Key for the Inngest Source API
    */
   constructor(
```

Comment:
While we don't necessarily need to worry about backcompat since this is pre-stable, we're going to need 2 different keys for sending (Source API key) and registering (Inngest API key). I'm definitely open to new ways that we could improve this, but `send()` wants a different API key which could be a challenge.

My gut is also saying that the constructor is heading towards being more overloaded, so perhaps just a single object argument may be the best and most flexible going forward.

Reply:
Great shout - I had missed the need for two keys. Cool - I'll add this!